### PR TITLE
[WIP] add allowed_address_pairs

### PR DESCRIPTION
**Pull request by roobert**
Opening this PR (as requested by @jtopjian) as a work in progress to support Allowed Address Pairs[1] now that gophercloud has support has been merged[2].

[1] http://specs.openstack.org/openstack/neutron-specs/specs/api/allowed_address_pairs.html
[2] https://github.com/rackspace/gophercloud/pull/488#issuecomment-182002690

### CODE REVIEW

**Comment by nathwill** [comment]
thanks for putting this together @roobert! super excited to see it land in terraform :) super important for us in how we're doing HA deployments on OpenStack.

**Comment by roobert** [comment]
Hi @nathwill, I've actually stopped working on this as we've moved away from OpenStack at work.

I while ago I was discussing this with @jtopjian via email as he's the author of the awesome openstack terraform provider.

Here was my last email to Joe:

```
Hi Joe,

I've been looking at this again today and have come across another problem.

I've been manually running neutron and nova commands to create what I would consider to be a "correct" HA setup with allowed-address-pairs and keepalived, I've then tried to convert this into something I can deploy using terraform.

The first issue I've come across is that I can't assign a floating-ip resource to a port. I'm somewhat fuzzy on the details but there seems to be a difference between how 'neutron' and 'nova' cli tools work. Whether they use the same API or not or just behave differently, I'm not sure but I'm guessing the terraform provider behaves in the same way 'nova' works. In order to associate a floating-ip to a port I have to use 'neutron' like so (after creating a port with terraform, nova, or neutron):

neutron floatingip-create --port-id <port> <floating network>

This creates a floating ip and assigns it at the same time and is the first part of the configuration. After the port is setup on the internal network it's then just a matter of updating each compute instances port with the ip address of the port with the associated floating-ip:

neutron port-update <compute instance port id> --allowed-address-pairs list=true type=dict ip_address=<ip of initially created port>

So basically there are two things that are currently not possible and I'm not sure how to go about implementing them:

1) need to be able to associate a port with a floating-ip
2) need to be able to modify compute instance ports that are automatically created when a compute instance is created

Would love to hear your thoughts on this..

Cheers,

Rob
```

He helpfully pointed out this resource (I must've missed!): https://www.terraform.io/docs/providers/openstack/r/networking_floatingip_v2.html

So I think it would be possible to finish this work, but unfortunately I've been reassigned to other projects now and won't have the time.

I hope this helps though!

**Comment by nathwill** [comment]
@roobert thanks for the heads up on this being abandoned, that means we're probably looking at other solutions like Heat for now.

as it seems you already discovered, floating ip's aren't really a replacement for allowed-address-pairs. 

we tend to use allowed-address-pairs for keepalived-managed internal VIPs, where we don't care to allocate external address space. but, they're also useful for building HA public services where neutron load-balancing is not available: for example, you can create an internal port (https://www.terraform.io/docs/providers/openstack/r/networking_port_v2.html), and then assign an external floating ip to that port (https://www.terraform.io/docs/providers/openstack/r/compute_floatingip_v2.html), but in order to be able to associate the internal port to specific server instances, you have to be able to add the internal port as an allowed-address-pair for the ports that are assigned to those nova instances, or neutron's address spoofing filters block the traffic when you add that floating internal port to the interface on your nova instance.

anyways, thanks for the start on this; we'll definitely revisit terraform at some point in the future, when the OpenStack support's hopefully more in line with our particular needs.

**Comment by toddnni** [comment]
We are interested in getting this. Thanks for posting the unfinished work, hopefully I have time to check this out at some point.

Currently we use a hackish workaround that uses local-exec
```
provisioner "local-exec" {
    command = "neutron port-update `neutron port-list |grep '\"${self.access_ip_v4}\"' | awk {'print $2'}` --allowed-address-pairs type=dict list=true ip_address=${var.ha_ip}"
}
```

**Comment by jtopjian** [comment]
#8257 implemented this. :)

**Comment by roobert** [comment]
Thanks to all who worked on this, it's suddenly become relevant again!
